[AMD] Add Concat op to AMDGPU dialect

plognjen opened the pull request:
The "concat" operation combines a list of source n-dimensional tensors into a single larger destination tensor.

All source tensors must have the same shape, element type, and encoding.
The concatenation dimension is inferred from the source and destination shapes provided by the user.
For example, two tensors of shape 64x128 can produce a destination shape of 128x128,
indicating concatenation along dimension 0; or 64x256, indicating concatenation along dimension 1.

Generally, source tensors passed as op arguments can be arranged into the resulting shape in multiple ways.
For example, given four tensors of shape 64x64:
  concat s0<64x64>, s1<64x64>, s2<64x64>, s3<64x64> -> <128x128>

They can be laid out in different configurations within the result tensor:
   1) s0 s1 
       s2 s3  

   2) s0 s2
        s1 s3

From a logical tensor perspective, the source tensors are treated as elements of a tensor of tensors.
In other words, the 1-D array of input tensors is conceptually reshaped into an n-D grid.
The semantics of this op assume a row-major order (or its n-D generalization),
meaning the fastest-varying dimension is filled first, and the slowest-varying dimension is filled last.
In the example above, this corresponds to layout 1).

The source and destination tensors must have identical linear layouts at the CTA tile level.
That is, all base vectors for input dimensions must match, except for the register input dimension.
The register basis must align on the subset that defines the logical tensor shape of a single CTA tile.

This ensures that the concatenation is a no-op, meaning no data rearrangement among threads is required
to assemble the destination tensor with the given shape and layout.
However, the order of CTA tiles within the layout does not need to match between source and destination layouts.
It is the responsibility of the op's lowering logic to handle this correctly.

This op is designed to work on logical tensors directly, avoiding the need for complex layout reinterpretation or reshaping.
For example, the `tt.join` operation only supports concatenation along the innermost dimension,
and requires that the resulting innermost dimension provide 2 elements per thread, distributed across registers.
In contrast, this `concat` op imposes no constraints on the concatenation dimension or the size of dimensions.

ravil-mobile commented:
Hi @plognjen, I did the rebase of our internal `refine-ops` branch with your changes on the top and I tested accuracy of the pure GEMM and the fwd-FA kernels. Everything is green!

ravil-mobile commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ConcatOpToLLVM.cpp:
```diff
@@ -0,0 +1,126 @@
+    llvm::SmallVector<SmallVector<Value>> unpackedSources;
+    unpackedSources.reserve(sources.size());
+
+    for (int i = 0; i < sources.size(); i++) {
```
`size_t i = 0` b/c `sources.size()` returns `size_t`

ravil-mobile commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -302,4 +303,159 @@ InThreadTransposeOp::deduceOutputLayout(ArrayRef<int64_t> shape,
+                          int limit = -1) {
+    int n = (limit < 0 ? srcBasis.size()
+                       : std::min<unsigned>(srcBasis.size(), limit));
+    for (int i = 0; i < n; ++i) {
```
`size_t i = 0`

ravil-mobile commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -302,4 +303,159 @@ InThreadTransposeOp::deduceOutputLayout(ArrayRef<int64_t> shape,
+  auto [laneSrc, laneDst] = getBases("lane");
+  auto [warpSrc, warpDst] = getBases("warp");
+
+  auto shapeCTA_src = mlir::triton::gpu::getShapePerCTATile(srcType);
+  auto shapeCTA_dst = mlir::triton::gpu::getShapePerCTATile(dstType);
+  if (shapeCTA_src != shapeCTA_dst)
+    return emitError() << "CTA tile shapes must match between source and "
+                          "destination tensors.";
```
`camelCase`

guacamoleo commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
```diff
@@ -119,6 +119,58 @@ def ExtractSliceOp : TT_AMDGPU_Op<"extract_slice", [Pure]> {
+    does not need to match between source and destination layouts.
+    It is the responsibility of the op's lowering logic to handle this correctly.
+
+    * sources: a list of the input tensors.
```
Didn't we decide that we also need an `order` parameter here to determine row vs col major and such?
I see we have getRepOrder() which tries to infer an order, or return a default order. But I thought you had proved that we need the user to specify an order.

plognjen replied:
I managed to infer this info by analyzing destination layout. I was previously only thinking in terms of analyzing source layouts, which is why I thought we can't do it. We can talk more offline about the details.

guacamoleo replied:
That's great!
Just want to verify that Triton would always be able to infer the needed information from all dst layout types (blocked, swizzled, mfma...).
E.g. what if the src layouts are tiny mfma_layout, and the dst layout is also small mfma_layout (smaller than shapePerCTATile), or any edge cases like this.

plognjen replied:
It would be able to infer information for any **distributed** layout, so it doesn't work for swizzled shared layout for example. This is because extract_slice and concat ops are created for slicing in VGPRs. 
As for the second part, any layout will always have some shape for shapePerCTATile. This can be smaller then what we originally thought of CTA tile while using named layouts, but new definition of CTA tile is something like: smallest logical **contiguous** subtensor that is owned by a workgroup (so all of the launched threads and waprs own some elements of this subtensor). This will always be defined, it just may not be 32x32 which we originally thought of smallest CTA tile for MFMA for example.

guacamoleo replied:
Okay, I understand. Since we make the user provide a dst layout, which may be a LinearLayout (or convertable to LL), then we have all the details of every desired bases vector, and their relative order.

antiagainst commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ConcatOpToLLVM.cpp:
```diff
@@ -0,0 +1,127 @@
+
+      // After determining which source tensor the current CTA tile belongs to,
+      // compute the index of this CTA tile within that source tensor,
+      // considering the source tensor includes multiple CTA tiles.
```
s/includes/may include/

antiagainst commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ConcatOpToLLVM.cpp:
```diff
@@ -0,0 +1,127 @@
+          mlir::LLVM::linearize(multiDimSrcCTAIdx, srcCTAShape, srcCTAOrder);
+      auto unpackedElements = unpackedSources[linearSrcIdx];
+
+      for (int j = 0; j < elemsPerThreadPerCTA; ++j)
```
Use `append` or `append_range` to avoid iterating elementwise?

plognjen commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
```diff
@@ -119,6 +119,57 @@ def ExtractSliceOp : TT_AMDGPU_Op<"extract_slice", [Pure]> {
+  let summary = "concat operation";
+  let description = [{
+    The "concat" operation combines a list of source n-dimensional tensors into a single larger destination tensor.
+    Its semantics is designed to closely resemble Python-level slicing and operate on logical tensors,
```
Yes, you are right, it's misleading, will remove it.

plognjen commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
```diff
@@ -119,6 +119,57 @@ def ExtractSliceOp : TT_AMDGPU_Op<"extract_slice", [Pure]> {
+    The "concat" operation combines a list of source n-dimensional tensors into a single larger destination tensor.
+    Its semantics is designed to closely resemble Python-level slicing and operate on logical tensors,
+    avoiding the need for complex layout reinterpretation or reshaping.
+    This approach aims to make TTGIR easier to read and maintain compared to using join operation.
```
Concat can also join along innermost. The dimension along which concat is concatenating tensors is inferred from source and destination shapes. This flexibility is one selling point for concat over join as well. I'll add this in documentation.

plognjen commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
```diff
@@ -119,6 +119,57 @@ def ExtractSliceOp : TT_AMDGPU_Op<"extract_slice", [Pure]> {
+    This approach aims to make TTGIR easier to read and maintain compared to using join operation.
+
+    All source tensors must have the same shape, element type, and encoding.
+    They are arranged in minor-to-major order, forming a logical tensor of source tensors with shape
+    (dstDims[0] / srcDims[0], ..., dstDims[n-1] / srcDims[n-1]). When linearized, this matches the order
+    of the source tensors provided as arguments.
```
Ah, I see that wording is confusing here. It says **tensor of source tensors**, so its a tensor that contains multiple source tensors.

plognjen commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
```diff
@@ -119,6 +119,57 @@ def ExtractSliceOp : TT_AMDGPU_Op<"extract_slice", [Pure]> {
+    (dstDims[0] / srcDims[0], ..., dstDims[n-1] / srcDims[n-1]). When linearized, this matches the order
+    of the source tensors provided as arguments.
+
+    The source and destination must have matching linear layouts at the CTA tile level.
```
This would also be a bit misleading. Source tensor can also have multiple CTA tiles which are in different order than the ones in destination tensor. That way, it doesn't appear that reg base vectors in destination layout are simply appended along some output dimension. The condition here is for lane and warp base vectors to be the same, and subset of reg base vectors that describe logical sub-tensor that has the shape of 1 CTA tile to match.

plognjen replied:
How about this wording:

    The source and destination tensors must have identical linear layouts at the CTA tile level.
    That is, all base vectors for input dimensions must match, except for the register input dimension.
    The register basis must align on the subset that defines the logical tensor shape of a single CTA tile.

plognjen commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -302,4 +303,159 @@ InThreadTransposeOp::deduceOutputLayout(ArrayRef<int64_t> shape,
+  if (sources.empty())
+    return emitError() << "At least one source tensor is required.";
+
+  auto srcType = dyn_cast<RankedTensorType>(sources.front().getType());
+  if (!srcType)
+    return emitError() << "Expected source type is RankedTensorType.";
+
+  auto dstType = dyn_cast<RankedTensorType>(result.getType());
+  if (!dstType)
+    return emitError() << "Expected destination type is RankedTensorType.";
```
I'm not sure, I thought it's not checked if you define custom verifier.

plognjen replied:
You are right about this, actually all above are checked by ODS.

plognjen commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -302,4 +303,159 @@ InThreadTransposeOp::deduceOutputLayout(ArrayRef<int64_t> shape,
+    return emitError()
+           << "Source and destination tensors must have the same rank.";
+
+  // 2) Compute CTA tile counts and validate shapes.
```
Good catch, this is a typo. Thanks!

plognjen commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ConcatOpToLLVM.cpp:
```diff
@@ -0,0 +1,127 @@
+
+namespace {
+
+template <typename T> unsigned getNumElements(const ArrayRef<T> shape) {
```
It seems that computeProduct only works for int64_t types, but some utilities I use return unsigned, such as getShapePerCTATile. I could cast before calling computeProduct, but this just seemed more convinient.

plognjen commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ConcatOpToLLVM.cpp:
```diff
@@ -0,0 +1,127 @@
+    unsigned elemsPerTile = ::getNumElements<unsigned>(shapePerCTATile);
+    unsigned numCTATiles = totalElems / elemsPerTile;
+
+    // Default order is major-to-minor.
```
yes, it should be fastest to slowest varying, I'll abandon minor-major terminology completely in this PR to avoid confusion.

plognjen commented on third_party/amd/lib/TritonAMDGPUToLLVM/Utility.cpp:
```diff
@@ -752,4 +752,44 @@ void addLocalLoadNoAliasScope(AliasAnalysisOpInterface llLoadOp) {
+  // Compute number of CTA tiles in a layout.
+  unsigned totalElems = layout.getTotalOutDimSize();
+  auto ctaShape = llEnc.getShapePerCTATile();
+  unsigned elemsPerCTA =
```
Same as the above comment, getShapePerCTATile returns unsigned. Would you prefer to cast vector before passing to computeProduct or leave it with std::accumulate? I'm fine with both approaches

plognjen commented on third_party/amd/lib/TritonAMDGPUToLLVM/Utility.cpp:
```diff
@@ -752,4 +752,44 @@ void addLocalLoadNoAliasScope(AliasAnalysisOpInterface llLoadOp) {
+  // dimension. Then, for each subsequent basis register (first element of
+  // some CTA tile), extract the next nonzero index to build the full dimension
+  // order.
+  unsigned totalPerThread =
```
yes, that is more precise.

plognjen commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -302,4 +303,159 @@ InThreadTransposeOp::deduceOutputLayout(ArrayRef<int64_t> shape,
+
+void ConcatOp::getCanonicalizationPatterns(mlir::RewritePatternSet &patterns,
+                                           mlir::MLIRContext *context) {
+  patterns.add<CanonicalizeConcatOpFromExtractSlice>(context);
```
That's fair. Do you want me to then remove it from this patch and add it in subsequent patch where I've rewritten extract_slice?

plognjen commented on test/Conversion/amd/invalid_concat_op.mlir:
```diff
@@ -0,0 +1,67 @@
+    tensor<32x64xf32, #blocked>,tensor<32x64xf32, #blocked>, tensor<32x64xf32, #blocked>, tensor<32x64xf32, #blocked>, tensor<32x64xf32, #blocked>, tensor<32x64xf32, #blocked>, tensor<32x64xf32, #blocked>, tensor<32x64xf32, #blocked> -> tensor<128x128xf32, #blocked1>
+    tt.return
+  }
+}
```
Done.

antiagainst reviewed (changes requested):
Thanks for addressing the comments; looks much better now. Also looping in @ThomasRaoux to take a look to see if the semantics is solid now regarding linear layout.

ThomasRaoux commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -22,6 +22,7 @@
  */
 
 #include "triton/Dialect/Triton/IR/Dialect.h"
+#include "../../../TritonAMDGPUToLLVM/Utility.h"
```
nit: this is a layering problem

plognjen replied:
I agree, we can probably address this in some subsequent refactoring patch.

antiagainst replied:
Given we don't really need those utility functions in other files yet I moved them into the .cpp file to resolve this for now with 7ed0275. When we need it later let's have a proper place for them @plognjen.

ThomasRaoux commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -302,4 +303,96 @@ InThreadTransposeOp::deduceOutputLayout(ArrayRef<int64_t> shape,
+  auto [laneSrc, laneDst] = getBases("lane");
+  auto [warpSrc, warpDst] = getBases("warp");
+
+  auto shapeCTASrc = mlir::triton::gpu::getShapePerCTATile(srcType);
+  auto shapeCTADst = mlir::triton::gpu::getShapePerCTATile(dstType);
+  if (shapeCTASrc != shapeCTADst)
+    return emitError() << "CTA tile shapes must match between source and "
+                          "destination tensors.";
+
+  unsigned numCTAs = 1;
+  for (int d = 0; d < rank; ++d)
+    numCTAs *= srcShape[d] / shapeCTASrc[d];
+  unsigned elemsPerThread =
+      triton::gpu::getTotalElemsPerThread(srcType) / numCTAs;
+  unsigned regCompareLen = std::log2(elemsPerThread);
+
+  auto compareBasis = [&](auto &srcBasis, auto &dstBasis, StringRef message,
+                          int limit = -1) {
+    int n = (limit < 0 ? srcBasis.size()
+                       : std::min<unsigned>(srcBasis.size(), limit));
+    for (size_t i = 0; i < n; ++i) {
+      if (srcBasis[i] != dstBasis[i]) {
+        emitError() << message;
+        return false;
+      }
+    }
+    return true;
+  };
```
I think that works but it seems like over limiting. How are you deciding on the dst encoding when creating the op?

Overall the concept of CTA tiles feels a bit unnecessary and maybe add more complexity than needed.
Did you look how the join/split op are working? I though we could do something similar

plognjen replied:
Currently, concat is used in passes only with named layouts, where the source and destination layouts match, which works well with the proposed generalization. However, if it’s applied directly to LLs, it becomes the user’s responsibility to ensure the destination layout is formed correctly.

Compared to the semantics of join/split, this approach offers greater flexibility in operand/result shapes and in selecting the concatenation dimension, reducing the need for reshapes or layout reinterpretation. The tradeoff is that we work with "CTA tiles" as the smallest tensor unit, sacrificing some generality to gain more flexibility and work on "logical" tensors.

ThomasRaoux commented:
> Thanks for addressing the comments; looks much better now. Also looping in @ThomasRaoux to take a look to see if the semantics is solid now regarding linear layout.

I think the semantic looks good and works well with linear layout. I do feel like the code could be generalized/simplified.
I am curious how this works with layout propagation and how you calculate the layout when creating the op

plognjen commented:
> > Thanks for addressing the comments; looks much better now. Also looping in @ThomasRaoux to take a look to see if the semantics is solid now regarding linear layout.
> 
> I think the semantic looks good and works well with linear layout. I do feel like the code could be generalized/simplified. I am curious how this works with layout propagation and how you calculate the layout when creating the op

As for layout propagation, at the moment we use refinement/slicing pass after all remove layout passes and major optimizations, so we don't really need layouts to propagate trough concat op. However if we want to do this, we need to add logic to remove layout conversion pass in order for it to be aware how to propagate trough concat, something similar as for join and split in 
[Attribute inferSrcEncoding(Operation *op, Attribute encoding) function](https://github.com/triton-lang/triton/blob/main/lib/Dialect/TritonGPU/Transforms/Utility.cpp#L360).
@antiagainst since propagation trough is not a high priority currently, should we add this logic in subsequent patch?

antiagainst commented:
This patch focusing on the op definition itself is okay to sequence the work. But we do need to figure out layout propagation and implement it properly given that this op can be useful outside of AMD backend too.
